README.md
# Honkai_Starrail_bot
星穹铁路资源查询

## 简介

这个插件帮助群员在QQ群内查询星穹铁路的资源位点。
其他功能正在开发中。欢迎PR


## **部署**
1. 在HoshinoBot/hoshino/modules 目录下拉取本项目
```
git clone https://github.com/Cathgao/Honkai_Starrail_bot
```

2. 在hoshino/config/\_\_bot\_\_.py文件中，MODULES_ON里添加 "Honkai_Starrail_bot"


## **使用**

群聊中发送 `#哪里有[资源名称]` ，如 `#哪里有普通战利品`

发送 `星铁资源列表` 获取可查询的资源名称

## 鸣谢 
[Genshin_Impact_bot](https://github.com/H-K-Y/Genshin_Impact_bot)


CHANGELOG.md
# 2022-9-2

* 修复cq注入bug @ [0360473](https://github.com/H-K-Y/Genshin_Impact_bot/commit/0360473)
* 更新2.6材料图片 @ [2fe32dc](https://github.com/H-K-Y/Genshin_Impact_bot/commit/2fe32dc)
* 修复角色图鉴报错问题 @ [421772d](https://github.com/H-K-Y/Genshin_Impact_bot/commit/421772d)
* 修复资源点查询 @ [a579856](https://github.com/H-K-Y/Genshin_Impact_bot/commit/a579856)

# 2022-3-30

* 更新素材图片 @ [3b75fda](https://github.com/H-K-Y/Genshin_Impact_bot/commit/3b75fda)
* 更新圣遗物图片 @ [8475787](https://github.com/H-K-Y/Genshin_Impact_bot/commit/8475787)
* 更新材料图片 @ [2fe32dc](https://github.com/H-K-Y/Genshin_Impact_bot/commit/2fe32dc)

# 2021-12-14

* 修复武器池概率和保底错误的问题 @ [0123c1d](https://github.com/H-K-Y/Genshin_Impact_bot/commit/0123c1d)
* 添加圣遗物华馆、海染 @ [1c6ca19](https://github.com/H-K-Y/Genshin_Impact_bot/commit/1c6ca19)
* 未开放卡池数据混入的问题 @ [d471227](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d471227)

# 2021-11-23

* 获取卡池信息时加入时间判断 @ [6131b39](https://github.com/H-K-Y/Genshin_Impact_bot/commit/6131b39)
* 修复官方卡池数据变动导致只能抽取常驻池的问题 @ [8fa46ec](https://github.com/H-K-Y/Genshin_Impact_bot/commit/8fa46ec)
* 修复卡池问题 @ [89f162b](https://github.com/H-K-Y/Genshin_Impact_bot/commit/89f162b)
* 修复了抽卡、材料获取、黄历在频道中报错的问题 @ [fd28f9e](https://github.com/H-K-Y/Genshin_Impact_bot/commit/fd28f9e)


### 2021-10-30

* 解决超时导致的图鉴图标更新失败问题 @ [542f48e](https://github.com/H-K-Y/Genshin_Impact_bot/commit/542f48e)

### 2021-10-10

* 加入未实装角色的图鉴查找机制 @ [d861dde](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d861dde)
* 调整图鉴图标 @ [d2a3445](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d2a3445)
* 修复地图BUG @ [4c4c0ab](https://github.com/H-K-Y/Genshin_Impact_bot/commit/4c4c0ab)

### 2021-9-27

* 抽卡现在可以自动更新了 @ [cd80eba](https://github.com/H-K-Y/Genshin_Impact_bot/commit/cd80eba)
* 加入新指令 更新星铁卡池 @ [cd80eba](https://github.com/H-K-Y/Genshin_Impact_bot/commit/cd80eba)

### 2021-8-16

* 项目移植到nonebot 2  @ [6c1f73b](https://github.com/H-K-Y/Genshin_Impact_bot/commit/6c1f73b)

### 2021-8-14

* 加入 [圣遗物评分](https://github.com/H-K-Y/Genshin_Impact_bot/issues/31) @ [9e5eabf](https://github.com/H-K-Y/Genshin_Impact_bot/commit/9e5eabf)
* 更新卡池 @ [8b68ef6](https://github.com/H-K-Y/Genshin_Impact_bot/commit/8b68ef6)

### 2021-7-27

* 对资源查询的代码逻辑进行调整，防止资源查询时出现killed @ [928e983](https://github.com/H-K-Y/Genshin_Impact_bot/commit/928e983)
* 加入新命令 更新星铁地图 用于爬取大地图文件 @ [91f5422](https://github.com/H-K-Y/Genshin_Impact_bot/commit/91f5422)

### 2021-6-2

* 加入抽签功能 @ [33a4712](https://github.com/H-K-Y/Genshin_Impact_bot/commit/33a4712)
* 更新README.md @ [33a4712](https://github.com/H-K-Y/Genshin_Impact_bot/commit/33a4712)
* 部分代码依照PEP标准进行格式化 @ [33a4712](https://github.com/H-K-Y/Genshin_Impact_bot/commit/33a4712)


### 2021-3-28

* 圣遗物收集修改冰本的圣遗物名称，增加岩本 @ [c949f00](https://github.com/H-K-Y/Genshin_Impact_bot/commit/c949f00)
* 加入每日素材提醒 @ [c949f00](https://github.com/H-K-Y/Genshin_Impact_bot/commit/c949f00)

### 2021-3-24

* 加入了星铁黄历，黄历的数据来源于 [可莉特调](https://genshin.pub/) @ [3c85818](https://github.com/H-K-Y/Genshin_Impact_bot/commit/3c85818)

### 2021-3-11

* 对抽卡的概率进行调整现在更接近游戏里的效果 @ [44c33d5](https://github.com/H-K-Y/Genshin_Impact_bot/commit/44c33d5)
* 对抽卡逻辑优化防止在非武器UP池抽出限定4星武器 @ [44c33d5](https://github.com/H-K-Y/Genshin_Impact_bot/commit/44c33d5)
* 修正千岩系列武器图标星级错误的问题 @ [d7fc42a](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d7fc42a)

### 2020-12-13

* 资源列表查询现在改为直接抓取米游社地图的数据，每天自动更新 @ [24374bb](https://github.com/H-K-Y/Genshin_Impact_bot/commit/24374bb)
* 修复查询资源列表消息过长导致发送失败的问题 @ [88c219b](https://github.com/H-K-Y/Genshin_Impact_bot/commit/88c219b)

### 2020-11-26

* 修复模拟抽卡武器UP池保底次数错误的问题 @ [2abd006](https://github.com/H-K-Y/Genshin_Impact_bot/commit/2abd006)

### 2020-11-20

* 加入圣遗物收集功能 @ [c267e44](https://github.com/H-K-Y/Genshin_Impact_bot/commit/c267e44)

### 2020-11-3

* 修复了找神瞳可以添加相同名称的编号导致报错的问题 @ [e38bb44](https://github.com/H-K-Y/Genshin_Impact_bot/commit/e38bb44)

### 2020-10-31

* 加入了查询资源位置的功能 @ [3811750](https://github.com/H-K-Y/Genshin_Impact_bot/commit/3811750)

### 2020-10-29

* 加入了找神瞳的功能 @ [521c290](https://github.com/H-K-Y/Genshin_Impact_bot/commit/521c290)

### 2020-10-26

* 加入了丘丘语翻译功能 @ [f551e04](https://github.com/H-K-Y/Genshin_Impact_bot/commit/f551e04)

### 2020-10-22

* 修复了4星概率写错了导致4星只能保底抽出的问题 @ [4476bed](https://github.com/H-K-Y/Genshin_Impact_bot/commit/4476bed)
* 加入了更多抽卡的统计结果，比如抽出最多的武器是啥，第一个4星和5星事多少抽出的 @ [f6d5cc6](https://github.com/H-K-Y/Genshin_Impact_bot/commit/f6d5cc6)

### 2020-10-21

* 在config.py加入了抽卡限制功能，限制每个人一天最多抽多少次 @ [bcdcba0](https://github.com/H-K-Y/Genshin_Impact_bot/commit/bcdcba0)
    感谢[corvo007](https://github.com/corvo007)提出的粪pr (粪pr是他自己说的
* 修复了第一次切换卡池后不能再切换卡池的问题 @ [2ea0ea2](https://github.com/H-K-Y/Genshin_Impact_bot/commit/2ea0ea2)
* 修复了查看UP信息时报错的问题 @ [2ea0ea2](https://github.com/H-K-Y/Genshin_Impact_bot/commit/2ea0ea2)

### 2020-10-20

* 重构了项目代码 @ [d446ce6](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d446ce6)
    ~~第二天就重写？，过几天怕是忘了写的啥了........~~
* 加入的武器UP池和常驻池 @ [d446ce6](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d446ce6)
* 加入了切换卡池的功能 @ [d446ce6](https://github.com/H-K-Y/Genshin_Impact_bot/commit/d446ce6)

### 2020-10-19

* 加入10连抽，90连抽和180连抽 @ [c2f43b4](https://github.com/H-K-Y/Genshin_Impact_bot/commit/c2f43b4)
* 加入抽卡发送图片功能 @ [02b4592](https://github.com/H-K-Y/Genshin_Impact_bot/commit/02b4592)
* 修复了Windows系统发送图片时可能出现的路径问题，图片全部改为base64发送 @ [000122c](https://github.com/H-K-Y/Genshin_Impact_bot/commit/000122c)

query_resource_points/__init__.py

from loguru import logger
from hoshino import Service
from .query_resource_points import get_resource_map_mes, get_resource_list_mes, init_point_list_and_map
import re

sv = Service("星铁资源查询")


# @sv.on_rex(r"(\w+)(?:在哪|在哪里|哪有|哪里有)")
@sv.on_rex(r"(#)(?:哪有|哪里有)(\w+)")
async def inquire_resource_points(bot, ev):
    resource_name = ev['plain_text']
    resource_name = re.findall("(?<=['有'])(.*?)$", resource_name)[0]
    if not resource_name :
        return
    mes_list = []
    mes,msg = await get_resource_map_mes(resource_name)
    if len(msg) :
        for res in msg:
            msg_data = f"{res['upname']}  {res['name']} \n [CQ:image,file={res['b64']}]"
            data = {
                "type": "node",
                "data": {
                    "name": "色图机器人",
                    "uin": "2854196310",
                    "content": msg_data
                }
            }
            mes_list.append(data)
    await bot.send(ev, mes, at_sender=True)
    if mes_list :
        await bot.send_group_forward_msg(group_id=ev['group_id'], messages=mes_list)


@sv.on_fullmatch('星铁资源列表')
async def inquire_resource_list(bot, ev):
    mes_list = []
    txt_list = get_resource_list_mes().split("\n")
    for txt in txt_list:
        data = {
            "type": "node",
            "data": {
                "name": "色图机器人",
                "uin": "2854196310",
                "content": txt
            }
        }
        mes_list.append(data)
    # await bot.send(ev, get_resource_list_mes(), at_sender=True)
    await bot.send_group_forward_msg(group_id=ev['group_id'], messages=mes_list)


@sv.on_fullmatch('刷新星铁资源列表')
async def refresh_resource_list(bot, ev):
    await init_point_list_and_map()
    await bot.send(ev, '刷新成功', at_sender=True)


@sv.on_fullmatch('更新星铁地图')
async def up_map_icon(bot, ev):
    await init_point_list_and_map()
    await bot.send(ev, '更新成功', at_sender=True)


query_resource_points/query_resource_points.py

from PIL import Image,ImageMath
from io import BytesIO
from loguru import logger
import json
import os
import time
import base64
import httpx
import asyncio
import shutil


MAPLIST_URL     = 'https://api-static.mihoyo.com/common/srmap/sr_map/v1/map/tree?' #所有地图名称
POINT_LIST_URL  = 'https://api-static.mihoyo.com/common/srmap/sr_map/v1/map/point/list?map_id=' #当前地图的标记，包含坐标与资源名称
LABEL_URL       = 'https://api-static.mihoyo.com/common/srmap/sr_map/v1/map/label/tree?map_id=' #所有资源汇总
MAP_URL         = "https://api-static.mihoyo.com/common/srmap/sr_map/v1/map/info?map_id=" #当前地图图片，包含区域名字
APP_LABEL       = 'app_sn=sr_map&lang=zh-cn' #请求标签，必须附加在url结尾

header = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/84.0.4147.105 Safari/537.36'

FILE_PATH = os.path.dirname(__file__)

MAP_PATH = os.path.join(FILE_PATH,"icon","map_icon.jpg")
Image.MAX_IMAGE_PIXELS = None


CENTER = None
ORIGIN = None

zoom = 0.5
resource_icon_offset = (-int(150*0.5*zoom),-int(150*zoom))
map_list = []
map_origin_list = []

data = {
    "all_resource_type":{
        # 这个字典保存所有资源类型，
        # "1": {
        #         "id": 1,
        #         "name": "传送点",
        #         "icon": "",
        #         "parent_id": 0,
        #         "depth": 1,
        #         "node_type": 1,
        #         "jump_type": 0,
        #         "jump_target_id": 0,
        #         "display_priority": 0,
        #         "children": []
        #     },
    },
    "can_query_type_list":{
        # 这个字典保存所有可以查询的资源类型名称和ID，这个字典只有名称和ID
        # 上边字典里"depth": 2的类型才可以查询，"depth": 1的是1级目录，不能查询
        # "七天神像":"2"
        # "风神瞳":"5"

    },
    "all_resource_point_list" :[
            # 这个列表保存当前地图资源点的数据
            # {
            #     "id": 2740,
            #     "label_id": 68,
            #     "x_pos": -1789,
            #     "y_pos": 2628,
            #     "author_name": "✟紫灵心✟",
            #     "ctime": "2020-10-29 10:41:21",
            #     "display_state": 1
            # },
    ],
    "date":"" #记录上次更新"all_resource_point_list"的日期
}

async def get_map_pointlist(map_num):
    async with httpx.AsyncClient() as client:
        resp = await client.get(url=POINT_LIST_URL + str(map_num) + "&"+APP_LABEL)
        if resp.status_code != 200:
            raise ValueError(f"获取地图资源失败，错误代码 {resp.status_code}")
        map_list = resp.content
        return resp.content
        

async def download_icon(url):
    # 下载图片，返回Image对象
    async with httpx.AsyncClient() as client:
        resp = await client.get(url=url)
        if resp.status_code != 200:
            raise ValueError(f"获取图片数据失败，错误代码 {resp.status_code}")
        icon = resp.content
        return Image.open(BytesIO(icon))

async def download_json(url):
    # 获取资源数据，返回 JSON
    async with httpx.AsyncClient() as client:
        resp = await client.get(url=url)
        if resp.status_code != 200:
            raise ValueError(f"获取资源点数据失败，错误代码 {resp.status_code}")
        return resp.json()


async def up_icon_image(sublist):
    # 检查是否有图标，没有图标下载保存到本地
    id = sublist["id"]
    icon_path = os.path.join(FILE_PATH,"icon",f"{id}.png")

    if not os.path.exists(icon_path):
        logger.info(f"正在更新资源图标 {id}")
        icon_url = sublist["icon"]
        icon = await download_icon(icon_url)
        icon = icon.resize((150, 150))

        box_alpha = Image.open(os.path.join(FILE_PATH,"icon","box_alpha.png")).getchannel("A")
        box = Image.open(os.path.join(FILE_PATH,"icon","box.png"))

        try:
            icon_alpha = icon.getchannel("A")
            icon_alpha = ImageMath.eval("convert(a*b/256, 'L')", a=icon_alpha, b=box_alpha)
        except ValueError:
            # 米游社的图有时候会没有alpha导致报错，这时候直接使用box_alpha当做alpha就行
            icon_alpha = box_alpha

        icon2 = Image.new("RGBA", (150, 150), "#00000000")
        icon2.paste(icon, (0, 0))

        bg = Image.new("RGBA", (150, 150), "#00000000")
        bg.paste(icon2, mask=icon_alpha)
        bg.paste(box, mask=box)

        with open(icon_path, "wb") as icon_file:
            bg.save(icon_file)


async def up_label_and_point_list():
    logger.info(f"正在更新资源点数据")
    label_data = await download_json(LABEL_URL + "3&" + APP_LABEL)
    for label in label_data["data"]["tree"]:
        data["all_resource_type"][str(label["id"])] = label
        for sublist in label["children"]:
            data["all_resource_type"][str(sublist["id"])] = sublist
            data["can_query_type_list"][sublist["name"]] = str(sublist["id"])
            await up_icon_image(sublist)
        label["children"] = []
    data["date"] = time.strftime("%d")
    
    for index, map in enumerate(map_list):
        label_data = await download_json(POINT_LIST_URL + str(map["id"]) + "&" + APP_LABEL)
        label_list = []
        for label in label_data["data"]["label_list"]:
            label_list += [label["id"]]
        map_list[index] = {"id": map["id"], "upname": map["upname"], "name":map["name"], "origin": map["origin"], "resource":label_list}
    data["date"] = time.strftime("%d")
    logger.info(f"资源点数据更新完成")

async def download_map(map_id):
    # 下载地图文件
    global map_origin_list
    map_info = await download_json(MAP_URL + str(map_id) + "&" + APP_LABEL)
    map_info = map_info["data"]["info"]["detail"]
    if(map_info == ""):
            return
    map_info = json.loads(map_info)
    map_origin = map_info["origin"]
    if not os.path.exists(os.path.join(FILE_PATH,"maps")):
        os.mkdir(os.path.join(FILE_PATH, "maps"))
    map_path = os.path.join(FILE_PATH,"maps",f"map_{map_id}.png")
    if not os.path.exists(map_path):
        logger.info(f"正在下载地图{map_id}图片")
        map_url = map_info['slices'][0][0]["url"]
        map = await download_icon(map_url)
        with open(map_path, "wb") as icon_file:
            map.save(icon_file)
        # map_origin_list += [{"id":map_id,"origin":map_origin}]
    for index, map in enumerate(map_list):
        if (map["id"] == map_id):
            map_list[index] = {"id": map["id"], "upname": map["upname"], "name": map["name"], "origin": map_origin}

    
async def up_map():
    logger.info(f"正在更新地图数据")
    map_list_info = await download_json(MAPLIST_URL + APP_LABEL)
    for maps_tree in map_list_info["data"]["tree"]:
        if("children" in maps_tree):
            for map in maps_tree["children"]:
                global map_list
                map_list += [{"id": map["id"], "upname": maps_tree["name"], "name":map["name"]}]
                await download_map(map["id"])
    logger.info(f"地图数据更新完成")


async def init_point_list_and_map():
    await up_map()
    await up_label_and_point_list()
    with open(os.path.join(FILE_PATH, f"map_list.json"), "w", encoding='utf8') as map_list_json:
        json.dump(map_list, map_list_json, ensure_ascii=False, indent=2)
    



# 初始化
loop = asyncio.get_event_loop()
loop.run_until_complete(init_point_list_and_map())




class Resource_map(object):

    def __init__(self,resource_name,map_id):
        global CENTER
        self.resource_id = str(data["can_query_type_list"][resource_name])
        map_path = os.path.join(FILE_PATH,"maps",f"map_{map_id}.png")
        self.map_id = map_id
        self.map_image = Image.open(map_path)
        self.map_size = self.map_image.size

        # 地图要要裁切的左上角和右下角坐标
        # 这里初始化为地图的大小
        self.x_start = self.map_size[0]
        self.y_start = self.map_size[1]
        CENTER = [self.map_size[0], self.map_size[1]]
        self.resource_icon = Image.open(self.get_icon_path())
        self.resource_icon = self.resource_icon.resize((int(150*zoom), int(150*zoom)))

        self.resource_xy_list = self.get_resource_point_list()

    def get_icon_path(self):
        # 检查有没有图标，有返回正确图标，没有返回默认图标
        icon_path = os.path.join(FILE_PATH, "icon", f"{self.resource_id}.png")

        if os.path.exists(icon_path):
            return icon_path
        else:
            return os.path.join(FILE_PATH, "icon", "0.png")

    def get_resource_point_list(self):
        temp_list = []
        for resource_point in data["all_resource_point_list"]:
            if str(resource_point["label_id"]) == self.resource_id:
                # 获取xy坐标
                x = resource_point["x_pos"]
                y = resource_point["y_pos"]
                temp_list.append((int(x), int(y)))
        return temp_list

    def paste(self):
        for x, y in self.resource_xy_list:
            # 把资源图片贴到地图上
            # 这时地图已经裁切过了，要以裁切后的地图左上角为中心再转换一次坐标
            x += ORIGIN[0]
            y += ORIGIN[1]
            self.map_image.paste(self.resource_icon, (x + resource_icon_offset[0], y + resource_icon_offset[1]), self.resource_icon)
            # self.map_image.paste(self.resource_icon,(x, y),self.resource_icon)

    def get_cq_cod(self):

        if not self.resource_xy_list:
            return "没有这个资源的信息"

        # self.crop()

        self.paste()

        bio = BytesIO()
        self.map_image.save(bio, format='JPEG')
        base64_str = 'base64://' + base64.b64encode(bio.getvalue()).decode()

        return f"[CQ:image,file={base64_str}]"

    def save_img(self):
        if not self.resource_xy_list:
            return "没有这个资源的信息"
        self.paste()
        self.map_image = self.map_image.convert('RGB')
        with open(os.path.join(FILE_PATH, "tmp",f"map_{self.map_id}_res{self.resource_id}.jpg"), "wb") as tmp_img:
            self.map_image.save(tmp_img, format='JPEG')
    
    def get_BIO(self):
        self.paste()
        bio = BytesIO()
        self.map_image = self.map_image.convert('RGB')
        self.map_image.save(bio, format='JPEG')
        return bio
         
    def get_resource_count(self):
        return len(self.resource_xy_list)

async def check_resource_on_map(resource_name):
    for map in map_list:
        current_map = await get_map_pointlist(map["id"])
        

async def get_resource_map_mes(name):
    count = 0
    mes = None
    img_list = []
    if data["date"] !=  time.strftime("%d"):
        await init_point_list_and_map()

    if not (name in data["can_query_type_list"]):
        mes = f"没有 {name} 这种资源。\n发送 星铁资源列表 查看所有资源名称"
        return mes,img_list
    
    try :
        shutil.rmtree(os.path.join(FILE_PATH, "tmp"))
    except:
        os.mkdir(os.path.join(FILE_PATH, "tmp"))
    else:
        os.mkdir(os.path.join(FILE_PATH, "tmp"))   
    resource_id = int(data["can_query_type_list"][name])
    for map in map_list:
        if (resource_id in map["resource"]):
            global ORIGIN
            map_id = map["id"]
            label_data = await download_json(f"{POINT_LIST_URL}{map_id}&{APP_LABEL}")
            data["all_resource_point_list"] = label_data["data"]["point_list"]
            ORIGIN = map["origin"]
            map_obj = Resource_map(name, map_id)
            base64_str = 'base64://' + base64.b64encode(map_obj.get_BIO().getvalue()).decode()
            img_list += [{"upname": map["upname"], "name":map["name"], "b64":base64_str}]
            count += map_obj.get_resource_count()
    mes = f"※ {name} 一共找到 {count} 个位置点\n※ 数据来源于米游社wiki"

    if not count:
        mes = f"没有找到 {name} 资源的位置，可能米游社wiki还没更新。"
        return mes,img_list
    
    return mes,img_list



def get_resource_list_mes():

    temp = {}

    for id in data["all_resource_type"].keys():
        # 先找1级目录
        if data["all_resource_type"][id]["depth"] == 1:
            temp[id] = []

    for id in data["all_resource_type"].keys():
        # 再找2级目录
        if data["all_resource_type"][id]["depth"] == 2:
            temp[str(data["all_resource_type"][id]["parent_id"])].append(id)

    mes = "当前资源列表如下：\n"

    for resource_type_id in temp.keys():

        if resource_type_id in ["1","12","50","51","95","131"]:
            # 在游戏里能查到的数据这里就不列举了，不然消息太长了
            continue

        mes += f"{data['all_resource_type'][resource_type_id]['name']}："
        for resource_id in temp[resource_type_id]:
            mes += f"{data['all_resource_type'][resource_id]['name']}，"
        mes += "\n"

    return mes
